Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -0,0 +1,1191 @@
+
+Source of this name: https://en.wikipedia.org/wiki/Quadratic_form#Generalization
+-/
+def polar (f : M → R) (x y : M) :=
+  f (x + y) - f x - f y
+#align quadratic_form.polar QuadraticForm.polar
+
+theorem polar_add (f g : M → R) (x y : M) : polar (f + g) x y = polar f x y + polar g x y := by
+  simp only [polar, Pi.add_apply]
+  abel
+#align quadratic_form.polar_add QuadraticForm.polar_add
+
+theorem polar_neg (f : M → R) (x y : M) : polar (-f) x y = -polar f x y := by
+  simp only [polar, Pi.neg_apply, sub_eq_add_neg, neg_add]
+#align quadratic_form.polar_neg QuadraticForm.polar_neg
+
+theorem polar_smul [Monoid S] [DistribMulAction S R] (f : M → R) (s : S) (x y : M) :
+    polar (s • f) x y = s • polar f x y := by simp only [polar, Pi.smul_apply, smul_sub]
+#align quadratic_form.polar_smul QuadraticForm.polar_smul
+
+theorem polar_comm (f : M → R) (x y : M) : polar f x y = polar f y x := by
+  rw [polar, polar, add_comm, sub_sub, sub_sub, add_comm (f x) (f y)]
+#align quadratic_form.polar_comm QuadraticForm.polar_comm
+
+/-- Auxiliary lemma to express bilinearity of `QuadraticForm.polar` without subtraction. -/
+theorem polar_add_left_iff {f : M → R} {x x' y : M} :
+    polar f (x + x') y = polar f x y + polar f x' y ↔
+      f (x + x' + y) + (f x + f x' + f y) = f (x + x') + f (x' + y) + f (y + x) := by
+  simp only [← add_assoc]
+  simp only [polar, sub_eq_iff_eq_add, eq_sub_iff_add_eq, sub_add_eq_add_sub, add_sub]
+  simp only [add_right_comm _ (f y) _, add_right_comm _ (f x') (f x)]
+  rw [add_comm y x, add_right_comm _ _ (f (x + y)), add_comm _ (f (x + y)),
+    add_right_comm (f (x + y)), add_left_inj]
+#align quadratic_form.polar_add_left_iff QuadraticForm.polar_add_left_iff
+
+theorem polar_comp {F : Type _} [Ring S] [AddMonoidHomClass F R S] (f : M → R) (g : F) (x y : M) :
+    polar (g ∘ f) x y = g (polar f x y) := by
+  simp only [polar, Pi.smul_apply, Function.comp_apply, map_sub]
+#align quadratic_form.polar_comp QuadraticForm.polar_comp
+
+end QuadraticForm
+
+end Polar
+
+/-- A quadratic form over a module.
+
+For a more familiar constructor when `R` is a ring, see `QuadraticForm.ofPolar`. -/
+structure QuadraticForm (R : Type u) (M : Type v) [Semiring R] [AddCommMonoid M] [Module R M] where
+  toFun : M → R
+  toFun_smul : ∀ (a : R) (x : M), toFun (a • x) = a * a * toFun x
+  exists_companion' : ∃ B : BilinForm R M, ∀ x y, toFun (x + y) = toFun x + toFun y + B x y
+#align quadratic_form QuadraticForm
+
+namespace QuadraticForm
+
+section FunLike
+
+variable [Semiring R] [AddCommMonoid M] [Module R M]
+
+variable {Q Q' : QuadraticForm R M}
+
+instance funLike : FunLike (QuadraticForm R M) M fun _ => R where
+  coe := toFun
+  coe_injective' x y h := by cases x; cases y; congr
+#align quadratic_form.fun_like QuadraticForm.funLike
+
+/-- Helper instance for when there's too many metavariables to apply
+`FunLike.hasCoeToFun` directly. -/
+instance : CoeFun (QuadraticForm R M) fun _ => M → R :=
+  ⟨FunLike.coe⟩
```

Comment:
I recall that we decided to only port the `FunLike` instances and skip the `CoeFun` instances (as also suggested by the comment in `LinearMap`). Did we change this at some point?